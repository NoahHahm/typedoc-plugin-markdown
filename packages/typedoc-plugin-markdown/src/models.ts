import { ReflectionKind, TypeDocOptionMap } from 'typedoc';

export interface TypedocPluginMarkdownOptions extends TypeDocOptionMap {
  anchorFormat: AnchorFormat;
  baseUrl: string;
  enableFrontmatter: boolean;
  entryDocument: string;
  enumMembersFormat: DataFormatStyle;
  flattenOutputFiles: boolean;
  frontmatterTags: string[];
  frontmatterGlobals: FrontmatterGlobals;
  frontmatterNamingConvention: FrontmatterNamingConvention;
  groupByKinds: boolean;
  hideBreadcrumbs: boolean;
  hideInPageTOC: boolean;
  hideKindTag: boolean;
  hidePageTitle: boolean;
  hideHierarchy: boolean;
  indentifiersAsCodeBlocks: boolean;
  indexPageTitle: string;
  kindsWithOwnFile: string | string[];
  namedAnchors: boolean;
  numberPrefixOutput: boolean;
  propertiesFormat: DataFormatStyle;
  TOCFormat: DataFormatStyle;
}

export interface TemplateMapping {
  directory: string | null;
  template: any;
  isLeaf: boolean;
  kind: ReflectionKind;
}

export type Collapse = 'object' | 'function' | 'all' | 'none';

export type DataFormatStyle = 'list' | 'table';

export type AnchorFormat = 'lowercase' | 'slug' | 'none';

export type FrontmatterGlobals =
  | string
  | Record<string, string | boolean | number | null>;

export type FrontmatterNamingConvention =
  | 'camelCase'
  | 'snakeCase'
  | 'kebabCase'
  | 'pascalCase';

export interface NavigationItem {
  title?: string;
  url?: string;
  children?: NavigationItem[];
}
