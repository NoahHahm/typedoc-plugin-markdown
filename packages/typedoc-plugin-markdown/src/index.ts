/**
 * Exposes the plugin API of the plugin
 */
export { load } from './plugin/bootstrap';
export { MarkdownRendererEvent } from './plugin/renderer';
export { MarkdownTheme, MarkdownThemeRenderContext } from './theme/definition';
export { NavigationItem } from './theme/models';
