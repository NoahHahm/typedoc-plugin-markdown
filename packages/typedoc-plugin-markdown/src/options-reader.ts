import { Options, OptionsReader } from 'typedoc';
import { TypedocPluginMarkdownOptions } from './models';

export class MarkdownPluginOptionsReader implements OptionsReader {
  priority = 900;
  options: Partial<TypedocPluginMarkdownOptions>;

  name = 'custom-options';

  constructor(options: Partial<TypedocPluginMarkdownOptions>) {
    this.options = options;
  }

  read(container: Options) {
    if (this.options) {
      Object.entries(this.options).forEach(([key, value]) => {
        container.setValue(key, value);
      });
    }
  }
}
