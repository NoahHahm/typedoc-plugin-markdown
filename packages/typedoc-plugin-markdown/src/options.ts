import { Application, Options, OptionsReader, ParameterType } from 'typedoc';

export function defineOptions(app: Application) {
  app.options.addReader(
    new (class implements OptionsReader {
      name = 'markdown-theme-options-reader';
      readonly order = 900;
      readonly supportsPackages = false;
      read(container: Options) {
        if (container.getValue('theme') === 'default') {
          container.setValue('theme', 'markdown');
        }
      }
    })(),
  );

  /**
   * file output options
   */
  app.options.addDeclaration({
    name: 'entryDocument',
    help: '[Markdown Plugin] The file name of the entry document.',
    type: ParameterType.String,
    defaultValue: 'README.md',
  });

  app.options.addDeclaration({
    name: 'flattenOutput',
    help: '[Markdown Plugin] Flatten output files without directories.',
    type: ParameterType.Boolean,
    defaultValue: false,
  });

  app.options.addDeclaration({
    name: 'numberPrefixOutput',
    help: '[Markdown Plugin] Prefixes docs and folders by number prefixes if applicable.',
    type: ParameterType.Boolean,
    defaultValue: false,
  });

  app.options.addDeclaration({
    name: 'kindsWithOwnFile',
    help: "[Markdown Plugin] Specifies which reflection kinds are contained in their own file. Values 'none', 'all' OR Array of ['class', 'interface', 'enum', 'function', 'variable', 'type']",
    type: ParameterType.String | ParameterType.Array,
    defaultValue: 'all',
  });

  app.options.addDeclaration({
    name: 'groupByKinds',
    help: '[Markdown Plugin] Groups reflection kinds by headings if applicable e.g Classes, Functions. If set to false all symbols will render on the same level. Defaults to `true`',
    type: ParameterType.Boolean,
    defaultValue: true,
  });

  /**
   * ui options
   */

  app.options.addDeclaration({
    name: 'hidePageTitle',
    help: '[Markdown Plugin] Do not print page title.',
    type: ParameterType.Boolean,
    defaultValue: false,
  });

  app.options.addDeclaration({
    name: 'hideKindTag',
    help: '[Markdown Plugin] Do not print the kind tag beneath the page title.',
    type: ParameterType.Boolean,
    defaultValue: false,
  });

  app.options.addDeclaration({
    name: 'hideBreadcrumbs',
    help: '[Markdown Plugin] Do not print breadcrumbs.',
    type: ParameterType.Boolean,
    defaultValue: false,
  });

  app.options.addDeclaration({
    name: 'hideInPageTOC',
    help: '[Markdown Plugin] Do not render in-page table of contents items.',
    type: ParameterType.Boolean,
    defaultValue: false,
  });

  app.options.addDeclaration({
    name: 'hideHierarchy',
    help: '[Markdown Plugin] Do not print reflection hierarchy.',
    type: ParameterType.Boolean,
    defaultValue: false,
  });

  app.options.addDeclaration({
    name: 'indexPageTitle',
    help: '[Markdown Plugin] The title of the main index / modules page.',
    type: ParameterType.String,
  });

  app.options.addDeclaration({
    name: 'indentifiersAsCodeBlocks',
    help: '[Markdown Plugin] Format signature and declaration identifiers in code blocks.',
    type: ParameterType.Boolean,
    defaultValue: false,
  });

  app.options.addDeclaration({
    name: 'propertiesFormat',
    help: '[Markdown Plugin] Specify the render style of properties.',
    type: ParameterType.String,
    defaultValue: 'List',
  });

  app.options.addDeclaration({
    name: 'enumMembersFormat',
    help: '[Markdown Plugin] Specify the render style of properties.',
    type: ParameterType.String,
    defaultValue: 'List',
  });

  app.options.addDeclaration({
    name: 'typeDeclarationFormat',
    help: '[Markdown Plugin] Specify the render style of type declarations.',
    type: ParameterType.String,
    defaultValue: 'List',
  });

  /**
   * Frontmatter options
   */
  app.options.addDeclaration({
    name: 'enableFrontmatter',
    help: '[Markdown Plugin] Prepend frontmatter to output.',
    type: ParameterType.Boolean,
    defaultValue: false,
  });

  app.options.addDeclaration({
    name: 'frontmatterTags',
    help: '[Markdown Plugin] Specify which file comment tags should be added to frontmatter.',
    type: ParameterType.Array,
    defaultValue: [],
  });

  app.options.addDeclaration({
    name: 'frontmatterGlobals',
    help: '[Markdown Plugin] Specify static variables to be added to all frontmatter.',
    type: ParameterType.Mixed,
    defaultValue: {},
  });

  app.options.addDeclaration({
    name: 'frontmatterNamingConvention',
    help: '[Markdown Plugin] Specify the naming convention of front matter variables.',
    type: ParameterType.String,
    defaultValue: 'camelCase',
    validate: (option) => {
      const availableValues = [
        'camelCase',
        'snakeCase',
        'kebabCase',
        'pascalCase',
      ];
      if (!availableValues.includes(option)) {
        throw new Error(
          `Unexpected value for frontmatterNamingConvention, the expected value is one of 'camelCase', 'snakeCase','kebabCase','pascalCase'`,
        );
      }
    },
  });

  /**
   * Utility options
   */
  app.options.addDeclaration({
    help: '[Markdown Plugin] Specifies the base url for internal link. If omitted all urls will be relative.',
    name: 'baseUrl',
    type: ParameterType.String,
  });

  app.options.addDeclaration({
    name: 'namedAnchors',
    help: '[Markdown Plugin] Use HTML named anchors as fragment identifiers for engines that do not automatically assign header ids. Should be set for Bitbucket Server docs.',
    type: ParameterType.Boolean,
    defaultValue: false,
  });

  app.options.addDeclaration({
    name: 'preserveAnchorCasing',
    help: '[Markdown Plugin] Preserve anchor casing when generating links.',
    type: ParameterType.Boolean,
    defaultValue: false,
  });
}
