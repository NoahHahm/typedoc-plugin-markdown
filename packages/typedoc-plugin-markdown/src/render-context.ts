import * as path from 'path';
import { Options, PageEvent, Reflection } from 'typedoc';
import { URL_PREFIX } from './support/constants';

/* start_imports */
import { breadcrumbs } from './resources/partials/breadcrumbs';
import { commentParts } from './resources/partials/comment.parts';
import { comment } from './resources/partials/comment';
import { footer } from './resources/partials/footer';
import { header } from './resources/partials/header';
import { hierarchy } from './resources/partials/hierarchy';
import { declarationMemberIdentifier } from './resources/partials/member.declaration.identifier';
import { declarationMemberName } from './resources/partials/member.declaration.name';
import { declarationMember } from './resources/partials/member.declaration';
import { memberIndex } from './resources/partials/member.index';
import { indexSignatureTitle } from './resources/partials/member.indexsignature.title';
import { inheritance } from './resources/partials/member.inheritance';
import { memberKindTag } from './resources/partials/member.kind-tag';
import { referenceMember } from './resources/partials/member.reference';
import { reflectionMember } from './resources/partials/member.reflection';
import { signatureMemberIdentifier } from './resources/partials/member.signature.identifier';
import { signatureMember } from './resources/partials/member.signature';
import { sources } from './resources/partials/member.sources';
import { memberTitle } from './resources/partials/member.title';
import { member } from './resources/partials/member';
import { typeDeclarationMember } from './resources/partials/member.type-declaration';
import { members } from './resources/partials/members';
import { navigation } from './resources/partials/navigation';
import { pageIndex } from './resources/partials/page.index';
import { pageTitle } from './resources/partials/page.title';
import { enumMembersTable } from './resources/partials/table.enum-members';
import { parametersTable } from './resources/partials/table.parameters';
import { propertiesTable } from './resources/partials/table.properties';
import { typeParametersTable } from './resources/partials/table.typeparameters';
import { arrayType } from './resources/partials/type.array';
import { conditionalType } from './resources/partials/type.conditional';
import { declarationType } from './resources/partials/type.declaration';
import { functionType } from './resources/partials/type.function';
import { indexAccessType } from './resources/partials/type.index-access';
import { inferredType } from './resources/partials/type.inferred';
import { intersectionType } from './resources/partials/type.intersection';
import { intrinsicType } from './resources/partials/type.intrinsic';
import { literalType } from './resources/partials/type.literal';
import { queryType } from './resources/partials/type.query';
import { referenceType } from './resources/partials/type.reference';
import { reflectionType } from './resources/partials/type.reflection';
import { someType } from './resources/partials/type.some';
import { tupleType } from './resources/partials/type.tuple';
import { typeOperatorType } from './resources/partials/type.type-operator';
import { unionType } from './resources/partials/type.union';
import { unknownType } from './resources/partials/type.unknown';
import { memberTemplate } from './resources/templates/member';
import { projectTemplate } from './resources/templates/project';
import { readmeTemplate } from './resources/templates/read-me';
import { reflectionTemplate } from './resources/templates/reflection';
/* end_imports */

function bind<F, L extends any[], R>(fn: (f: F, ...a: L) => R, first: F) {
  return (...r: L) => fn(first, ...r);
}

export class MarkdownThemeRenderContext {
  constructor(
    public page: PageEvent<Reflection> | null,
    public options: Options,
  ) {}

  urlTo = (reflection: Reflection) => {
    return this.relativeURL(reflection.url);
  };

  relativeURL = (url: string | undefined) => {
    if (!url) {
      return null;
    }
    if (URL_PREFIX.test(url)) {
      return url;
    } else {
      if (this.options.getValue('baseUrl')) {
        return this.options.getValue('baseUrl') + url.replace(/\\/g, '/');
      }

      const relative = path.relative(
        path.dirname(this.page?.url || '.'),
        path.dirname(url),
      );

      return this.parseUrl(
        path.join(relative, path.basename(url)).replace(/\\/g, '/'),
      );
    }
  };

  parseUrl(url: string) {
    return encodeURI(url);
  }

  /* start_resources */

  // templates
  memberTemplate = bind(memberTemplate, this);
  projectTemplate = bind(projectTemplate, this);
  readmeTemplate = bind(readmeTemplate, this);
  reflectionTemplate = bind(reflectionTemplate, this);

  // partials
  breadcrumbs = bind(breadcrumbs, this);
  commentParts = bind(commentParts, this);
  comment = bind(comment, this);
  footer = bind(footer, this);
  header = bind(header, this);
  hierarchy = bind(hierarchy, this);
  declarationMemberIdentifier = bind(declarationMemberIdentifier, this);
  declarationMemberName = bind(declarationMemberName, this);
  declarationMember = bind(declarationMember, this);
  memberIndex = bind(memberIndex, this);
  indexSignatureTitle = bind(indexSignatureTitle, this);
  inheritance = bind(inheritance, this);
  memberKindTag = bind(memberKindTag, this);
  referenceMember = bind(referenceMember, this);
  reflectionMember = bind(reflectionMember, this);
  signatureMemberIdentifier = bind(signatureMemberIdentifier, this);
  signatureMember = bind(signatureMember, this);
  sources = bind(sources, this);
  memberTitle = bind(memberTitle, this);
  member = bind(member, this);
  typeDeclarationMember = bind(typeDeclarationMember, this);
  members = bind(members, this);
  navigation = bind(navigation, this);
  pageIndex = bind(pageIndex, this);
  pageTitle = bind(pageTitle, this);
  enumMembersTable = bind(enumMembersTable, this);
  parametersTable = bind(parametersTable, this);
  propertiesTable = bind(propertiesTable, this);
  typeParametersTable = bind(typeParametersTable, this);
  arrayType = bind(arrayType, this);
  conditionalType = bind(conditionalType, this);
  declarationType = bind(declarationType, this);
  functionType = bind(functionType, this);
  indexAccessType = bind(indexAccessType, this);
  inferredType = bind(inferredType, this);
  intersectionType = bind(intersectionType, this);
  intrinsicType = bind(intrinsicType, this);
  literalType = bind(literalType, this);
  queryType = bind(queryType, this);
  referenceType = bind(referenceType, this);
  reflectionType = bind(reflectionType, this);
  someType = bind(someType, this);
  tupleType = bind(tupleType, this);
  typeOperatorType = bind(typeOperatorType, this);
  unionType = bind(unionType, this);
  unknownType = bind(unknownType, this);
  /* end_resources */
}
