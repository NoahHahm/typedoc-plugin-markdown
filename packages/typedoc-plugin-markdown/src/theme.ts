import * as prettier from 'prettier';
import {
  PageEvent,
  ProjectReflection,
  Reflection,
  RenderTemplate,
  Renderer,
  Theme,
} from 'typedoc';
import { MarkdownThemeConverterContext } from './theme-converter-context';
import { MarkdownThemeRenderContext } from './theme-render-context';

export class MarkdownTheme extends Theme {
  private _renderContext: MarkdownThemeRenderContext;
  private _converterContext: MarkdownThemeConverterContext;
  private _prettierOptions: prettier.Options | null;

  constructor(renderer: Renderer) {
    super(renderer);

    this.listenTo(this.owner, {
      [PageEvent.BEGIN]: this.onBeginPage,
    });
  }

  getRenderContext() {
    if (!this._renderContext) {
      this._renderContext = new MarkdownThemeRenderContext(
        this,
        this.application.options,
      );
    }
    return this._renderContext;
  }

  getConverterContext() {
    if (!this._converterContext) {
      this._converterContext = new MarkdownThemeConverterContext(
        this,
        this.application.options,
      );
    }
    return this._converterContext;
  }

  getPrettierOptions() {
    if (!this._prettierOptions) {
      this._prettierOptions = this.resolvePrettierOptions();
    }
    return this._prettierOptions;
  }

  resolvePrettierOptions() {
    const prettierFile = prettier.resolveConfigFile.sync();
    return prettierFile ? prettier.resolveConfig.sync(prettierFile) : null;
  }

  render(
    page: PageEvent<Reflection>,
    template: RenderTemplate<PageEvent<Reflection>>,
  ): string {
    return prettier.format(template(page) as string, {
      ...this.getPrettierOptions(),
      parser: 'markdown',
    });
  }

  getUrls(project: ProjectReflection) {
    return this.getConverterContext().getUrls(project);
  }

  getNavigation(project: ProjectReflection) {
    return this.getConverterContext().getNavigation(project);
  }

  protected onBeginPage(page: PageEvent) {
    this.getRenderContext().activeLocation = page.url;
  }
}
