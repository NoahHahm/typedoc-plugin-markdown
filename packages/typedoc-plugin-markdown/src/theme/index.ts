/**
 *
 * The in-built custom theme and context theme definitions that the plugin initiates.
 *
 * @module
 */

export * from './render-context';
export * from './theme';
