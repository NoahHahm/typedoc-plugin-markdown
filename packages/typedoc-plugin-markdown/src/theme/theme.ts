import {
  DeclarationReflection,
  ProjectReflection,
  Reflection,
  ReflectionKind,
  RenderTemplate,
  Renderer,
  Theme,
} from 'typedoc';
import { MarkdownPageEvent } from '../plugin/events';
import { OutputFileStrategy } from '../plugin/options/custom-maps';
import { UrlMapping } from '../plugin/url-mapping';
import { slugify } from '../support/utils';
import { NavigationItem, TemplateMapping } from './models';
import { NavigationContext } from './navigation-context';
import { MarkdownThemeRenderContext } from './render-context';
import { UrlsContext } from './urls-context';

/**
 * This is in-built MarkdownTheme which extends TypeDocs Theme class.
 * This follows the implementation of TypeDoc's [DefaultTheme](https://typedoc.org/api/classes/DefaultThemeRender.html).
 *
 * The {@link render } and {@link getUrls} methods is where the work happens.
 */
export class MarkdownTheme extends Theme {
  /**
   *
   * @param renderer The TypeDoc renderer instance the theme is attached to.
   */
  constructor(renderer: Renderer) {
    super(renderer);
  }

  getRenderContext(pageEvent: MarkdownPageEvent<Reflection> | null) {
    return new MarkdownThemeRenderContext(pageEvent, this.application.options);
  }

  getNavigationContext() {
    return new NavigationContext(this, this.application.options.getRawValues());
  }

  getUrlsContext() {
    return new UrlsContext(this, this.application.options.getRawValues());
  }

  readmeTemplate = (pageEvent: MarkdownPageEvent<ProjectReflection>) => {
    return this.getRenderContext(pageEvent).readmeTemplate(pageEvent);
  };

  projectTemplate = (pageEvent: MarkdownPageEvent<ProjectReflection>) => {
    return this.getRenderContext(pageEvent).projectTemplate(pageEvent);
  };

  reflectionTemplate = (
    pageEvent: MarkdownPageEvent<DeclarationReflection>,
  ) => {
    return this.getRenderContext(pageEvent).reflectionTemplate(pageEvent);
  };

  memberTemplate = (pageEvent: MarkdownPageEvent<DeclarationReflection>) => {
    return this.getRenderContext(pageEvent).memberTemplate(pageEvent);
  };

  /**
   * Renders a template and page model to a string.
   */
  render(
    page: MarkdownPageEvent<Reflection>,
    template: RenderTemplate<MarkdownPageEvent<Reflection>>,
  ) {
    return template(page) as string;
  }

  getUrls(project: ProjectReflection): UrlMapping[] {
    return this.getUrlsContext().getUrls(project);
  }

  getNavigation(project: ProjectReflection): NavigationItem[] {
    return this.getNavigationContext().getNavigation(project);
  }

  /**
   * Returns the template mapping for a given reflection kind
   * @param kind
   */
  getTemplateMapping(kind: ReflectionKind): TemplateMapping {
    const outputFileStrategy = this.application.options.getValue(
      'outputFileStrategy',
    ) as OutputFileStrategy;

    const getDirectoryName = (reflectionKind: ReflectionKind) => {
      const pluralString = ReflectionKind.pluralString(reflectionKind);
      return slugify(pluralString).toLowerCase();
    };

    const membersWithOwnFile = this.application.options.getValue(
      'membersWithOwnFile',
    ) as string[];

    const mappings = {
      [ReflectionKind.Module]: {
        template: this.reflectionTemplate,
        directory: null,
        kind: ReflectionKind.Module,
      },
      [ReflectionKind.Namespace]: {
        template: this.reflectionTemplate,
        directory: getDirectoryName(ReflectionKind.Namespace),
        kind: ReflectionKind.Namespace,
      },
    };

    if (
      outputFileStrategy === OutputFileStrategy.Members &&
      membersWithOwnFile?.includes(ReflectionKind[ReflectionKind.Class])
    ) {
      mappings[ReflectionKind.Class] = {
        template: this.reflectionTemplate,
        directory: getDirectoryName(ReflectionKind.Class),
        kind: ReflectionKind.Class,
      };
    }
    if (
      outputFileStrategy === OutputFileStrategy.Members &&
      membersWithOwnFile?.includes(ReflectionKind[ReflectionKind.Interface])
    ) {
      mappings[ReflectionKind.Interface] = {
        isLeaf: false,
        template: this.reflectionTemplate,
        directory: getDirectoryName(ReflectionKind.Interface),
        kind: ReflectionKind.Interface,
      };
    }
    if (
      outputFileStrategy === OutputFileStrategy.Members &&
      membersWithOwnFile?.includes(ReflectionKind[ReflectionKind.Enum])
    ) {
      mappings[ReflectionKind.Enum] = {
        template: this.reflectionTemplate,
        directory: getDirectoryName(ReflectionKind.Enum),
        kind: ReflectionKind.Enum,
      };
    }
    if (
      outputFileStrategy === OutputFileStrategy.Members &&
      membersWithOwnFile?.includes(ReflectionKind[ReflectionKind.Function])
    ) {
      mappings[ReflectionKind.Function] = {
        template: this.memberTemplate,
        directory: getDirectoryName(ReflectionKind.Function),
        kind: ReflectionKind.Function,
      };
    }
    if (
      outputFileStrategy === OutputFileStrategy.Members &&
      membersWithOwnFile?.includes(ReflectionKind[ReflectionKind.TypeAlias])
    ) {
      mappings[ReflectionKind.TypeAlias] = {
        template: this.memberTemplate,
        directory: getDirectoryName(ReflectionKind.TypeAlias),
        kind: ReflectionKind.TypeAlias,
      };
    }
    if (
      outputFileStrategy === OutputFileStrategy.Members &&
      membersWithOwnFile?.includes(ReflectionKind[ReflectionKind.Variable])
    ) {
      mappings[ReflectionKind.Variable] = {
        template: this.memberTemplate,
        directory: getDirectoryName(ReflectionKind.Variable),
        kind: ReflectionKind.Variable,
      };
    }
    return mappings[kind];
  }
}
