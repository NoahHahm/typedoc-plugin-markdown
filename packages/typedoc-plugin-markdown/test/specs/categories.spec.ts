import { ProjectReflection } from 'typedoc';
import { MarkdownThemeRenderContext } from '../../src/theme';

describe(`Categories:`, () => {
  let project: ProjectReflection;
  let context: MarkdownThemeRenderContext;

  const baseOptions = {
    hideKindPrefix: true,
    disableSources: true,
    outputFileStrategy: 'modules',
  };

  const stubbedPartials = [
    'declarationMember',
    'reflectionMember',
    'signatureMember',
  ];

  describe(`(excludeGroups=false)`, () => {
    beforeAll(async () => {
      ({ project, context } = await global.bootstrap(['categories.ts'], {
        options: {
          ...baseOptions,
          excludeGroups: false,
        },
        stubPartials: stubbedPartials,
      }));
    });

    test(`should render category members'`, () => {
      expect(context.members(project, 2)).toMatchSnapshot();
    });

    test(`should render category TOC'`, () => {
      expect(context.reflectionIndex(project, 2)).toMatchSnapshot();
    });
  });

  describe(`(excludeGroups=true)`, () => {
    beforeAll(async () => {
      ({ project, context } = await global.bootstrap(['categories.ts'], {
        options: {
          ...baseOptions,
          excludeGroups: true,
        },
        stubPartials: stubbedPartials,
      }));
    });

    test(`should render category members'`, () => {
      expect(context.members(project, 2)).toMatchSnapshot();
    });

    test(`should render category TOC'`, () => {
      expect(context.reflectionIndex(project, 2)).toMatchSnapshot();
    });
  });

  describe(`(categorizeByGroup=false)`, () => {
    beforeAll(async () => {
      ({ project, context } = await global.bootstrap(['categories.ts'], {
        options: {
          ...baseOptions,
          categorizeByGroup: false,
        },
        stubPartials: stubbedPartials,
      }));
    });

    test(`should render category members'`, () => {
      expect(context.members(project, 2)).toMatchSnapshot();
    });

    test(`should render category TOC'`, () => {
      expect(context.reflectionIndex(project, 2)).toMatchSnapshot();
    });
  });
});
