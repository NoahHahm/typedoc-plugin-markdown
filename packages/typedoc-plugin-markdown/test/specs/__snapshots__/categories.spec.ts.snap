// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Categories: (categorizeByGroup=false) should render category TOC' 1`] = `
"## Index

### CategoryA

- [CategoryInterfaceA1](API.md#categoryinterfacea1)
- [CategoryInterfaceA2](API.md#categoryinterfacea2)
- [categoryVariableA1](API.md#categoryvariablea1)
- [categoryVariableA2](API.md#categoryvariablea2)
- [categoryFunctionA1](API.md#categoryfunctiona1)
- [categoryFunctionA2](API.md#categoryfunctiona2)
- [categoryFunctionA3](API.md#categoryfunctiona3)

### CategoryB

- [categoryVariableB1](API.md#categoryvariableb1)
- [categoryFunctionB2](API.md#categoryfunctionb2)

### Other

- [OtherInterfaceA](API.md#otherinterfacea)
- [otherVariableA](API.md#othervariablea)
- [otherVariableB](API.md#othervariableb)
- [otherFunctionA](API.md#otherfunctiona)
- [otherFunctionB](API.md#otherfunctionb)
"
`;

exports[`Categories: (categorizeByGroup=false) should render category members' 1`] = `
"## CategoryA

### CategoryInterfaceA1

[partial: reflectionMember]



---

### CategoryInterfaceA2

[partial: reflectionMember]



---

### categoryVariableA1

[partial: declarationMember]



---

### categoryVariableA2

[partial: declarationMember]



---

### categoryFunctionA1()

[partial: signatureMember]



---

### categoryFunctionA2()

[partial: signatureMember]



---

### categoryFunctionA3()

[partial: signatureMember]

## CategoryB

### categoryVariableB1

[partial: declarationMember]



---

### categoryFunctionB2()

[partial: signatureMember]

## Other

### OtherInterfaceA

[partial: reflectionMember]



---

### otherVariableA

[partial: declarationMember]



---

### otherVariableB

[partial: declarationMember]



---

### otherFunctionA()

[partial: signatureMember]



---

### otherFunctionB()

[partial: signatureMember]"
`;

exports[`Categories: (excludeGroups=false) should render category TOC' 1`] = `
"## Index

### Interfaces

#### CategoryA

- [CategoryInterfaceA1](API.md#categoryinterfacea1)
- [CategoryInterfaceA2](API.md#categoryinterfacea2)
- [OtherInterfaceA](API.md#otherinterfacea)

#### Other

- [CategoryInterfaceA1](API.md#categoryinterfacea1)
- [CategoryInterfaceA2](API.md#categoryinterfacea2)
- [OtherInterfaceA](API.md#otherinterfacea)

### Variables

#### CategoryA

- [categoryVariableA1](API.md#categoryvariablea1)
- [categoryVariableA2](API.md#categoryvariablea2)
- [categoryVariableB1](API.md#categoryvariableb1)
- [otherVariableA](API.md#othervariablea)
- [otherVariableB](API.md#othervariableb)

#### CategoryB

- [categoryVariableA1](API.md#categoryvariablea1)
- [categoryVariableA2](API.md#categoryvariablea2)
- [categoryVariableB1](API.md#categoryvariableb1)
- [otherVariableA](API.md#othervariablea)
- [otherVariableB](API.md#othervariableb)

#### Other

- [categoryVariableA1](API.md#categoryvariablea1)
- [categoryVariableA2](API.md#categoryvariablea2)
- [categoryVariableB1](API.md#categoryvariableb1)
- [otherVariableA](API.md#othervariablea)
- [otherVariableB](API.md#othervariableb)

### Functions

#### CategoryA

- [categoryFunctionA1](API.md#categoryfunctiona1)
- [categoryFunctionA2](API.md#categoryfunctiona2)
- [categoryFunctionA3](API.md#categoryfunctiona3)
- [categoryFunctionB2](API.md#categoryfunctionb2)
- [otherFunctionA](API.md#otherfunctiona)
- [otherFunctionB](API.md#otherfunctionb)

#### CategoryB

- [categoryFunctionA1](API.md#categoryfunctiona1)
- [categoryFunctionA2](API.md#categoryfunctiona2)
- [categoryFunctionA3](API.md#categoryfunctiona3)
- [categoryFunctionB2](API.md#categoryfunctionb2)
- [otherFunctionA](API.md#otherfunctiona)
- [otherFunctionB](API.md#otherfunctionb)

#### Other

- [categoryFunctionA1](API.md#categoryfunctiona1)
- [categoryFunctionA2](API.md#categoryfunctiona2)
- [categoryFunctionA3](API.md#categoryfunctiona3)
- [categoryFunctionB2](API.md#categoryfunctionb2)
- [otherFunctionA](API.md#otherfunctiona)
- [otherFunctionB](API.md#otherfunctionb)
"
`;

exports[`Categories: (excludeGroups=false) should render category members' 1`] = `
"## Interfaces

### CategoryA

#### CategoryInterfaceA1

[partial: reflectionMember]



---

#### CategoryInterfaceA2

[partial: reflectionMember]

### Other

#### OtherInterfaceA

[partial: reflectionMember]

## Variables

### CategoryA

#### categoryVariableA1

[partial: declarationMember]



---

#### categoryVariableA2

[partial: declarationMember]

### CategoryB

#### categoryVariableB1

[partial: declarationMember]

### Other

#### otherVariableA

[partial: declarationMember]



---

#### otherVariableB

[partial: declarationMember]

## Functions

### CategoryA

#### categoryFunctionA1()

[partial: signatureMember]



---

#### categoryFunctionA2()

[partial: signatureMember]



---

#### categoryFunctionA3()

[partial: signatureMember]

### CategoryB

#### categoryFunctionB2()

[partial: signatureMember]

### Other

#### otherFunctionA()

[partial: signatureMember]



---

#### otherFunctionB()

[partial: signatureMember]"
`;

exports[`Categories: (excludeGroups=true) should render category TOC' 1`] = `
"## Index

### CategoryA

- [CategoryInterfaceA1](API.md#categoryinterfacea1)
- [CategoryInterfaceA2](API.md#categoryinterfacea2)
- [categoryVariableA1](API.md#categoryvariablea1)
- [categoryVariableA2](API.md#categoryvariablea2)
- [categoryFunctionA1](API.md#categoryfunctiona1)
- [categoryFunctionA2](API.md#categoryfunctiona2)
- [categoryFunctionA3](API.md#categoryfunctiona3)

### CategoryB

- [categoryVariableB1](API.md#categoryvariableb1)
- [categoryFunctionB2](API.md#categoryfunctionb2)

### Other

- [OtherInterfaceA](API.md#otherinterfacea)
- [otherVariableA](API.md#othervariablea)
- [otherVariableB](API.md#othervariableb)
- [otherFunctionA](API.md#otherfunctiona)
- [otherFunctionB](API.md#otherfunctionb)
"
`;

exports[`Categories: (excludeGroups=true) should render category members' 1`] = `
"## CategoryA

### CategoryInterfaceA1

[partial: reflectionMember]



---

### CategoryInterfaceA2

[partial: reflectionMember]



---

### categoryVariableA1

[partial: declarationMember]



---

### categoryVariableA2

[partial: declarationMember]



---

### categoryFunctionA1()

[partial: signatureMember]



---

### categoryFunctionA2()

[partial: signatureMember]



---

### categoryFunctionA3()

[partial: signatureMember]

## CategoryB

### categoryVariableB1

[partial: declarationMember]



---

### categoryFunctionB2()

[partial: signatureMember]

## Other

### OtherInterfaceA

[partial: reflectionMember]



---

### otherVariableA

[partial: declarationMember]



---

### otherVariableB

[partial: declarationMember]



---

### otherFunctionA()

[partial: signatureMember]



---

### otherFunctionB()

[partial: signatureMember]"
`;
